packages/atproto_client/models/string_formats.py

```diff
@@ -0,0 +1,218 @@
+import re
+from datetime import datetime
+from typing import Callable, Mapping, Set, Union
+from urllib.parse import urlparse
+
+from atproto_core.nsid import validate_nsid as atproto_core_validate_nsid
+from pydantic import BeforeValidator, Field, ValidationInfo
+from pydantic_core import core_schema
+from typing_extensions import Annotated, Literal
+
+_OPT_IN_KEY: Literal['strict_string_format'] = 'strict_string_format'
+
+# constants
+MAX_HANDLE_LENGTH: int = 253
+MAX_NSID_LENGTH: int = 317
+MAX_RECORD_KEY_LENGTH: int = 512
+MAX_URI_LENGTH: int = 8 * 1024
+MIN_CID_LENGTH: int = 8
+TID_LENGTH: int = 13
+INVALID_RECORD_KEYS: Set[str] = {'.', '..'}
+MAX_DID_LENGTH: int = 2048  # Method-specific identifier max length
+MAX_AT_URI_LENGTH: int = 8 * 1024
+
+# patterns
+DOMAIN_RE = re.compile(r'^([a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z][a-zA-Z0-9-]*[a-zA-Z]$')
+DID_RE = re.compile(
+    r'^did:'  # Required prefix
+    r'[a-z]+:'  # method-name (lowercase only)
+    r'[a-zA-Z0-9._%-]{1,2048}'  # method-specific-id with length limit
+    r'(?<!:)$'  # Cannot end with colon
+)
+LANG_RE = re.compile(r'^(i|[a-z]{2,3})(-[A-Za-z0-9-]+)?$')
+RKEY_RE = re.compile(r'^[A-Za-z0-9._:~-]{1,512}$')
+TID_RE = re.compile(rf'^[2-7a-z]{{{TID_LENGTH}}}$')
+CID_RE = re.compile(r'^[A-Za-z0-9+]{8,}$')
+AT_URI_RE = re.compile(
+    r'^at://'  # Must start with at://
+    r'('  # Authority group start
+    # For DIDs: Only allowed chars are letters, numbers, period, hyphen, and percent
+    r'did:[a-z]+:[a-zA-Z0-9.-]+'  # Notice removed underscore from allowed chars
+    r'|'  # or
+    # Handle: require 2+ segments, TLD can't start with digit
+    r'[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?'  # First segment
+    r'(\.[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*'  # Middle segments
+    r'\.[a-zA-Z][a-zA-Z0-9-]*'  # TLD must start with letter
+    r')'  # Authority group end
+    r'(?:'  # Optional path group
+    r'/[a-z][a-zA-Z0-9-]*(\.[a-z][a-zA-Z0-9-])+'  # NSID
+    r'(?:/[A-Za-z0-9._:~-]+)?'  # Optional record key
+    r')?$'
+)
+
+
+def only_validate_if_strict(validate_fn: core_schema.WithInfoValidatorFunction) -> Callable:
+    """Skip validation if not opting into strict validation."""
+
+    def wrapper(v: str, info: ValidationInfo) -> str:
+        if info and isinstance(info.context, Mapping) and info.context.get(_OPT_IN_KEY, False):
+            return validate_fn(v, info)
+        return v
+
+    return wrapper
+
+
+@only_validate_if_strict
+def validate_handle(v: str, info: ValidationInfo) -> str:
```

Comment:
as i can see the `info` arg is not used? could be names like `_` instead

Reply:
i updated `only_validate_if_strict` to selectively pass `info` - let me know what you think!

Reply:
to be honest i am not sure about it. since there is no use case for use of 'info' the logic to support it is complicated and uses inspect module :( could we just drop this arg? without checking needs or not

Reply:
makes sense! i was thinking its likely one would want access to `info` later, but no need to get ahead of ourselves!

I've simplified it in [bad5512](https://github.com/MarshalX/atproto/pull/451/commits/bad5512e3bc8b8c7996d81ecdc56f7876aab5ea4)